File: sdks/go/entitlements.go
package settlekit

import (
	"context"
	"net/http"
)

// ListEntitlementsOptions filters ListEntitlements. CustomerID is required by
// the API; the others are optional.
type ListEntitlementsOptions struct {
	CustomerID string
	ProductID  string
	ActiveOnly bool
}

// ListEntitlements lists a customer's entitlements. GET /v1/entitlements.
func (c *Client) ListEntitlements(ctx context.Context, opts ListEntitlementsOptions) ([]Entitlement, error) {
	q := newQuery()
	q.add("customerId", opts.CustomerID)
	q.add("productId", opts.ProductID)
	if opts.ActiveOnly {
		q.add("activeOnly", "true")
	}
	var out []Entitlement
	if err := c.do(ctx, http.MethodGet, "/v1/entitlements"+q.encode(), nil, &out); err != nil {
		return nil, err
	}
	return out, nil
}

// GetEntitlement fetches an entitlement by id. GET /v1/entitlements/:id.
func (c *Client) GetEntitlement(ctx context.Context, id string) (*Entitlement, error) {
	var out Entitlement
	if err := c.do(ctx, http.MethodGet, "/v1/entitlements/"+id, nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// VerifyEntitlementInput is the body for VerifyEntitlement
// (POST /v1/entitlements/verify). Provide a feature, a product, and/or required
// credits to check.
type VerifyEntitlementInput struct {
	CustomerID      string `json:"customerId"`
	ProductID       string `json:"productId,omitempty"`
	Feature         string `json:"feature,omitempty"`
	RequiredCredits int    `json:"requiredCredits,omitempty"`
}

// VerifyEntitlementResult is the verify decision. Fields beyond Granted vary by
// entitlement type and are preserved in Raw for forward compatibility.
type VerifyEntitlementResult struct {
	Granted          bool           `json:"granted"`
	Reason           string         `json:"reason,omitempty"`
	CreditsRemaining *int           `json:"creditsRemaining,omitempty"`
	Raw              map[string]any `json:"-"`
}

// VerifyEntitlement checks whether a customer has access (feature / credits /
// product). This is the hot-path SDK call. POST /v1/entitlements/verify.
func (c *Client) VerifyEntitlement(ctx context.Context, in VerifyEntitlementInput) (*VerifyEntitlementResult, error) {
	var out VerifyEntitlementResult
	if err := c.do(ctx, http.MethodPost, "/v1/entitlements/verify", in, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// SpendCreditsInput is the body for SpendCredits
// (POST /v1/entitlements/spend-credits).
type SpendCreditsInput struct {
	CustomerID string `json:"customerId"`
	ProductID  string `json:"productId"`
	Amount     int    `json:"amount"`
}

// SpendCredits debits credits from a product entitlement.
// POST /v1/entitlements/spend-credits.
func (c *Client) SpendCredits(ctx context.Context, in SpendCreditsInput) (*Entitlement, error) {
	var out Entitlement
	if err := c.do(ctx, http.MethodPost, "/v1/entitlements/spend-credits", in, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// RevokeEntitlement revokes an entitlement with an optional reason.
// POST /v1/entitlements/:id/revoke.
func (c *Client) RevokeEntitlement(ctx context.Context, id, reason string) (*Entitlement, error) {
	body := map[string]any{}
	if reason != "" {
		body["reason"] = reason
	}
	var out Entitlement
	if err := c.do(ctx, http.MethodPost, "/v1/entitlements/"+id+"/revoke", body, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

File: apps/api/src/routes/entitlements.ts
/**
 * Entitlement routes (plan §14, §4).
 *
 * The universal access layer. Lists a customer's entitlements, verifies access
 * (feature flag / credits / product) via `EntitlementService.verify`, spends
 * credits, and revokes. The hot-path verify endpoint is what the SDK calls.
 */
import { Hono } from "hono";
import type { Context } from "hono";
import { z } from "zod";
import { notFound } from "@settlekit/common";
import type { AppEnv } from "../context.js";
import { data } from "../http/respond.js";
import { parseBody, validate } from "../http/validate.js";

const verifySchema = z
  .object({
    customerId: z.string().min(1),
    productId: z.string().optional(),
    feature: z.string().optional(),
    requiredCredits: z.number().int().positive().optional(),
  })
  .strict();

const spendSchema = z.object({
  customerId: z.string().min(1),
  productId: z.string().min(1),
  amount: z.number().int().positive(),
});

const revokeSchema = z.object({
  reason: z.string().min(1).default("revoked via API"),
});

export function entitlementRoutes(): Hono<AppEnv> {
  const app = new Hono<AppEnv>();

  // List a customer's entitlements (optionally active-only / per product).
  app.get("/", async (c) => {
    const ctx = c.get("ctx");
    const customerId = c.req.query("customerId");
    if (!customerId) throw notFound("customerId query param is required");
    const activeOnly = c.req.query("activeOnly") === "true";
    const productId = c.req.query("productId");
    const list = await ctx.entitlementRepo.listByCustomer(customerId, {
      activeOnly,
      ...(productId !== undefined ? { productId } : {}),
    });
    return data(c, list);
  });

  // Verify access (feature / credits / product).
  app.post("/verify", async (c) => {
    const body = await parseBody(c, verifySchema);
    const result = await c.get("ctx").entitlements.verify(body);
    return data(c, result);
  });

  // Spend credits against a product entitlement.
  app.post("/spend-credits", async (c) => {
    const ctx = c.get("ctx");
    const body = await parseBody(c, spendSchema);
    const updated = await ctx.entitlements.spendCredits(
      body.customerId,
      body.productId,
      body.amount,
    );
    return data(c, updated);
  });

  app.get("/:id", async (c) => {
    const ent = await c.get("ctx").entitlementRepo.findById(c.req.param("id"));
    if (!ent) throw notFound("entitlement not found", { id: c.req.param("id") });
    return data(c, ent);
  });

  // Revoke an entitlement.
  app.post("/:id/revoke", async (c) => {
    const ctx = c.get("ctx");
    const body = validate(revokeSchema, await safeJson(c));
    const revoked = await ctx.entitlements.revoke(c.req.param("id"), body.reason);
    return data(c, revoked);
  });

  return app;
}

/** Read a JSON body that may be empty, returning `{}` when absent. */
async function safeJson(c: Context<AppEnv>): Promise<unknown> {
  try {
    return await c.req.json();
  } catch {
    return {};
  }
}
